simulation/chart.py

```diff
@@ -1,44 +1,147 @@
 
+from enum import Enum
 from subprocess import Popen
-from typing import Optional, Tuple
+from typing import Any, Dict, NamedTuple, Optional, Tuple, TypedDict, Union
 
 import ffmpeg
 import pyqtgraph as pg
 from pyqtgraph.Qt import QtCore, QtGui
 
 
-def create_layout(
-    size: Tuple[int, int],
-    title: str,
-    show: bool,
-    antialias: Optional[bool] = None,
-    foreground=None,
-) -> pg.GraphicsLayoutWidget:
-    """Initialize application and layout container
+class RGBAColor(NamedTuple):
+    """An RGB color value represents RED, GREEN, and BLUE light sources
+    with alpha channel- which specifies the opacity for a color."""
+
+    R: int
+    """Red [0-255]"""
+    G: int
+    """Green [0-255]"""
+    B: int
+    """Blue [0-255]"""
+    A: Optional[int]
+    """Alpha [0-255]"""
+
+
+class NamedColor(Enum):
+    "Single-character string representing a predefined color"
+    r = "r"
+    """Red"""
+    g = "g"
+    """Green"""
+    b = "b"
+    """Blue"""
+    c = "c"
+    """Cyan"""
+    m = "m"
+    """Magenta"""
+    y = "ky"
+    """Yellow"""
+    k = "k"
+    """Key (black)"""
+    w = "w"
+    """White"""
+
+
+class IndexedColor(NamedTuple):
+    """Color from a single index. Useful for stepping through a predefined list of colors."""
+
+    index: int
+    hues: int
+
+
+Color = Union[RGBAColor, NamedColor, IndexedColor, int, float, str]
+"""Color based on model, predefined values, single greyscale value (0.0 - 1.0) , color index or string representing color (“#RGB”,“#RGBA”,“#RRGGBB”,“#RRGGBBAA”)."""
+
+
+def _make_color(color: Color) -> Any:
+    """Converts color value to QColor"""
+    if isinstance(color, RGBAColor):
+        if color.A is None:
+            return pg.mkColor((color.R, color.G, color.B))
+        else:
+            return pg.mkColor((color.R, color.G, color.B, color.A))
+    elif isinstance(color, NamedColor):
+        return pg.mkColor(color.value)
+    elif isinstance(color, IndexedColor):
+        return pg.mkColor((color.index, color.hues))
+    else:
+        return pg.mkColor(color)
+
+
+class PenStyle(Enum):
+    """Pen style for drawing lines"""
+
+    # NoPen                    = QtCore.Qt.PenStyle.NoPen
+    SolidLine = QtCore.Qt.PenStyle.SolidLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashLine = QtCore.Qt.PenStyle.DashLine  # pyright:ignore [reportGeneralTypeIssues]
+    DotLine = QtCore.Qt.PenStyle.DotLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashDotLine = (
+        QtCore.Qt.PenStyle.DashDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    DashDotDotLine = (
+        QtCore.Qt.PenStyle.DashDotDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    # CustomDashLine           = QtCore.Qt.PenStyle.CustomDashLine
+    # MPenStyle                = QtCore.Qt.PenStyle.MPenStyle
+
+
+def _make_pen(
+    color: Optional[Color] = None,
+    width: Optional[float] = None,
+    style: Optional[PenStyle] = None,
+) -> Any:
```

Comment:
Unfortunately pyqtgraph is a wrapper for different base modules (PySide,PyQt6,etc..) and method has no specific type information such as QtPen
https://pyqtgraph.readthedocs.io/en/latest/api_reference/functions.html#pyqtgraph.mkPen

Reply:
Is it possible to run a few tests and check what the output type is? Even if it's undocumented.

Reply:
If not, I might still get rid of the `Any` type. The point of type hints in Python is to convey useful information for people developing downstream of you. A type hint of `Any` doesn't really add info. Also, I'm quite certain that it isn't a true `Any` in this case. For example, could this return an `int` or `string`? If not, it's not really an `Any`. It's probably some sort of `Union`

Reply:
pyqtgraph is loading module base on existance so it can be PySide.QtGui.QPen, PyQt5.QtGui.QPen....
for our dependency it is PySide.QtGui.QPen (debug).
those functions as internal to module level so I believe adding a hard dependency to base module should be avoided if possible.

Reply:
Is there a base class across all `QPen`?

Reply:
Unfortunately no, PySide has some base class called shiboken but PyQt5 has none.

Reply:
And this is not something you could `Union`?

Reply:
As the backing module is loaded at runtime, you want me to hard code a union on possible values. This means 4 more hard dependency, though our aim is to decouple them.

Reply:
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/26da0d9cb098f741ba3c5fa2a9a1f2b3170437bc